### PR TITLE
WIP - Integrate folding with IVC

### PR DESCRIPTION
**Pull request by dannywillems**

### CODE REVIEW

File: ivc/tests/add.rs [hunk 1]
```diff
@@ -96,10 +109,27 @@ pub fn test_simple_add() {
...
+    // The total number of columns in our circuit.
+    const N_COL_TOTAL: usize = 3 + N_WIT_IVC;
+
     // Folding Witness
     #[derive(Clone, Debug, PartialEq, Eq, Hash)]
     pub struct PlonkishWitness {
-        pub witness: GenericWitness<3, Evaluations<Fp, Radix2EvaluationDomain<Fp>>>,
+        pub witness: GenericWitness<N_COL_TOTAL, Evaluations<Fp, Radix2EvaluationDomain<Fp>>>,
+        // This does not have to be part of the witness... can be a static precompiled object.
+        pub fixed_selectors: Vec<Evaluations<Fp, Radix2EvaluationDomain<Fp>>>,
```
Comment (dannywillems) [inline ivc/tests/add.rs:132]: This should go in the instance then.

File: folding/src/expressions.rs [hunk 2]
```diff
@@ -602,6 +602,42 @@ impl<C: FoldingConfig> FoldingCompatibleExpr<C> {
             _ => panic!("unsupported"),
         }
     }
+
+    /// Maps variable (column index) in expression using the `mapper`
+    /// function. Can be used to modify (remap) the indexing of
+    /// columns after the expression is built.
+    pub fn map_variable(
```
Comment (dannywillems) [inline folding/src/expressions.rs:609]: This is cherry-picked [here](https://github.com/o1-labs/proof-systems/pull/2337)

File: ivc/src/ivc/columns.rs [hunk 3]
```diff
@@ -38,6 +38,11 @@ pub type IVCPoseidonColumn = PoseidonColumn<IVC_POSEIDON_STATE_SIZE, IVC_POSEIDO
 ///
 ///```text
 ///
+///         Inputs:
```
Comment (dannywillems) [inline ivc/src/ivc/columns.rs:41]: The following comments are cherry-picked in https://github.com/o1-labs/proof-systems/pull/2338

File: ivc/src/ivc/columns.rs [hunk 4]
```diff
@@ -311,101 +325,108 @@ impl ColumnIndexer for IVCColumn {
...
+                IVCColumn::Block6ConstR => Column::Relation(0),
+                IVCColumn::Block6ULeft => Column::Relation(1),
+                IVCColumn::Block6UOutput => Column::Relation(2),
 
-            IVCColumn::Block6ConstR => Column::Relation(0),
-            IVCColumn::Block6ULeft => Column::Relation(1),
-            IVCColumn::Block6UOutput => Column::Relation(2),
+                _ => panic!("Column selection not possible."),
+            })
+            .add_rel_offset(1),
```
Comment (dannywillems) [inline ivc/src/ivc/columns.rs:408]: Ah, the offset is added here.

File: poly-commitment/src/commitment.rs [hunk 5]
```diff
@@ -777,13 +777,16 @@ where
...
+                // our domain is d1, we will commit to d1,
+                // contsidering only every 8th element.
                 let s = (plnm.domain().size / domain.size) as usize;
                 let v: Vec<_> = (0..(domain.size())).map(|i| plnm.evals[s * i]).collect();
                 commit_evaluations(&v, basis)
             }
             std::cmp::Ordering::Equal => commit_evaluations(&plnm.evals, basis),
             std::cmp::Ordering::Greater => {
-                panic!("desired commitment domain size greater than evaluations' domain size")
+                panic!("desired commitment domain size ({}) greater than evaluations' domain size ({}):", domain.size, plnm.domain().size)
```
Comment (dannywillems) [inline poly-commitment/src/commitment.rs:789]: Cherry-picked https://github.com/o1-labs/proof-systems/pull/2344

File: ivc/src/ivc/interpreter.rs [hunk 6]
```diff
@@ -940,34 +1040,48 @@ where
...
+    env.set_assert_mapper(Box::new(move |x| fold_iteration.clone() * s5.clone() * x));
     constrain_u(env);
 
     env.set_assert_mapper(Box::new(move |x| x));
 }
 
 /// Instantiates the IVC circuit for folding. L is relaxed (folded)
 /// instance, and R is strict (new) instance that is being relaxed at
-/// this step. `N_COL_TOTAL` is the total number of columnsfor IVC + APP.
+/// this step. `N_COL_TOTAL` is the total number of columns for IVC + APP.
```
Comment (dannywillems) [inline ivc/src/ivc/interpreter.rs:1075]: Added [here](https://github.com/o1-labs/proof-systems/pull/2345)

File: ivc/src/ivc/interpreter.rs [hunk 7]
```diff
@@ -940,34 +1038,48 @@ where
     // the constraints that are created in the block block_num will have
     // the form selector(block_num)*C(X) and not just C(X).
 
+    let fold_iteration = env.read_column(IVCColumn::FoldIteration);
```
Comment (dannywillems) [inline ivc/src/ivc/interpreter.rs:1041]: See https://github.com/o1-labs/proof-systems/pull/2372/files

File: ivc/src/ivc/interpreter.rs [hunk 8]
```diff
@@ -991,22 +1104,61 @@ pub fn ivc_circuit<F, Ff, Env, PParams, const N_COL_TOTAL: usize, const N_CHALS:
...
         t_terms,
     );
-    process_challenges::<_, _, N_COL_TOTAL, N_CHALS>(env, hash_r, &chal_l, r);
-    process_u::<_, _, N_COL_TOTAL>(env, u_l, r);
+    process_challenges::<_, _, N_COL_TOTAL, N_CHALS>(env, fold_iteration, hash_r, &chal_l, r);
+    process_u::<_, _, N_COL_TOTAL>(env, fold_iteration, u_l, r);
+}
+
+/// Base case IVC circuit, completely turned off.
+pub fn ivc_circuit_base_case<F, Env, const N_COL_TOTAL: usize, const N_CHALS: usize>(
```
Comment (dannywillems) [inline ivc/src/ivc/interpreter.rs:1144]: See https://github.com/o1-labs/proof-systems/pull/2372/files

File: ivc/src/ivc/interpreter.rs [hunk 9]
```diff
@@ -923,6 +1017,10 @@ where
         let sel = env.read_column(IVCColumn::BlockSel(i));
         env.assert_zero(sel.clone() * (sel.clone() - Env::constant(F::one())));
     }
+
+    // For now iteration is only 0 or 1
```
Comment (dannywillems) [inline ivc/src/ivc/interpreter.rs:1021]: Not adding it in https://github.com/o1-labs/proof-systems/pull/2372/files. We don't want this constraint as it won't be the case that fold_iteration will be 1 or 0.

**Comment by volhovm** [comment]
This was merged in a PR train culminating with https://github.com/o1-labs/proof-systems/pull/2386

**Pull request closed without merging**